### PR TITLE
Resolve native return-type from native union-type in short ternary

### LINKED ISSUE #8089: Infer native return type from ternary
# Feature Request

I would expect one of the rules out of `SetList::TYPE_DECLARATION` would infer a native return type in https://getrector.com/demo/0e4a125c-bfa3-41fa-9b96-2af05c0f10a0

I guess it would be the job for `ReturnTypeFromStrictNativeCallRector` ?

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/rectorphp/rector/issues/8089

### CODE REVIEW

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 1]
```diff
@@ -95,7 +96,8 @@ public function refactorWithScope(Node $node, Scope $scope): ?Node
 
         $ternary = $return->expr;
 
-        $nativeTernaryType = $scope->getNativeType($ternary);
+        $returnScope = $return->expr->getAttribute(AttributeKey::SCOPE);
```
Comment (staabm) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php:99]: TIL when we need the scope from a node instead of the one passed into the rector rule

File: rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_union_void.inc [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+namespace Rector\Tests\TypeDeclaration\Rector\Class_\ReturnTypeFromStrictTernaryRector\Fixture;
+
+class SkipUnionVoid
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector/Fixture/skip_union_void.inc:5]: test covers the edge case mention in https://github.com/rectorphp/rector/issues/8089#issuecomment-1651415299

**Comment by staabm** [comment]
should be good to go

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 3]
```diff
@@ -95,7 +97,12 @@ public function refactorWithScope(Node $node, Scope $scope): ?Node
...
+        if ($returnScope === null) {
+            throw new ShouldNotHappenException();
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php]: return null is better here, to avoid overlapped reprint node that scope may noy yet exists, that skipped at RectifiedAnalyzer.

Comment (staabm) [reply rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php]: :+1:

done

File: rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php [hunk 4]
```diff
@@ -14,8 +14,10 @@
 use PHPStan\Type\MixedType;
 use PHPStan\Type\TypeCombinator;
 use PHPStan\Type\UnionType;
+use Rector\Core\Exception\ShouldNotHappenException;
```
Comment (samsonasik) [inline rules/TypeDeclaration/Rector/Class_/ReturnTypeFromStrictTernaryRector.php]:
```suggestion
```

**Review by samsonasik — APPROVED** [review]

**Comment by samsonasik** [comment]
Thank you @staabm
